### ISSUE TITLE
Able to build qtbase, but not qt5

### ISSUE THREAD

**Issue by rickvanosch** [mxe/mxe#2777]
Good evening,

I'm trying to build a docker image with mxe to cross compile a QT project on ubuntu for windows.

Attached you'll find the dockerfile I use. Currently it is installing qtbase which works fine, but if you replace that with qt5 or append qtcharts qtsvg(which i need for my project) it can no longer build the docker image. I get errors like:

`Failed to build package qtactiveqt for target i686-w64-mingw32.static!
------------------------------------------------------------
make[3]: *** [Makefile:94: sub-tools-make_first-ordered] Error 2
make[3]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2/src'
make[2]: *** [Makefile:53: sub-src-make_first] Error 2
make[2]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2'
make[1]: *** [Makefile:881: build-only-qtactiveqt_i686-w64-mingw32.static] Error 2
make[1]: Leaving directory '/opt/mxe'
real	0m57.175s
user	2m11.388s
sys	0m12.410s
------------------------------------------------------------
[log]      /opt/mxe/log/qtactiveqt_i686-w64-mingw32.static

make: *** [Makefile:869: /opt/mxe/usr/i686-w64-mingw32.static/installed/qtactiveqt] Error 1
make: *** Waiting for unfinished jobs....
[done]        qtsensors              i686-w64-mingw32.static                                  14348 KiB      1m5.926s
[done]        qtxmlpatterns          i686-w64-mingw32.static                                  367924 KiB     4m34.925s
[done]        qtscript               i686-w64-mingw32.static                                  53084 KiB      4m18.718s
The command '/bin/sh -c cd /opt &&     apt-get update &&     apt-get install -y cmake &&     git clone https://github.com/mxe/mxe.git &&    cd /opt/mxe && git pull &&     NPROC=$(($(nproc)+4)) &&     make --jobs=4 JOBS=4 MXE_TARGETS='i686-w64-mingw32.static' qt5 &&     ln -s /opt/mxe/usr/bin/i686-w64-mingw32.static-qmake-qt5 /usr/bin/qmake &&     qmake --version &&     apt -y autoremove &&     apt -y autoclean &&     apt -y clean &&     exit 0' returned a non-zero code: 2

I hope someone can help me get this working! If there's any other info required, feel free to ask.

Kind regards
[Dockerfile.txt](https://github.com/mxe/mxe/files/8127811/Dockerfile.txt)
`

**Comment by MarcoBorrini99** [comment]
Without full log it's difficult to say anything.

**Comment by rickvanosch** [comment]
> Without full log it's difficult to say anything.

Thanks @MarcoBorrini99 for your reply. I'm relatively new to docker, I will try to figure out how to access the logs

**Comment by rickvanosch** [comment]
I managed to get some logs, hope it helps anything:

`main.cpp: In function 'bool dllInstall(const QString&, bool)':
main.cpp:197:33: warning: cast between incompatible function types from 'FARPROC' {aka 'int (__attribute__((stdcall)) *)()'} to 'DllInstallProc' {aka 'long int (__attribute__((stdcall)) *)(int, const wchar_t*)'} [-Wcast-function-type]
  197 |     DllInstallProc DllInstall = reinterpret_cast<DllInstallProc>(GetProcAddress(hdll, "DllInstall"));
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.cpp: In function 'HRESULT dumpIdl(const QString&, const QString&, const QString&)':
main.cpp:276:31: warning: cast between incompatible function types from 'FARPROC' {aka 'int (__attribute__((stdcall)) *)()'} to 'DumpIDLProc' {aka 'long int (__attribute__((stdcall)) *)(const QString&, const QString&)'} [-Wcast-function-type]
  276 |         DumpIDLProc DumpIDL = reinterpret_cast<DumpIDLProc>(GetProcAddress(hdll, "DumpIDL"));
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
i686-w64-mingw32.static-g++ -Wl,-s -Wl,-subsystem,console -Wl,--gc-sections -o ../../../bin/idc.exe .obj/release/main.o .obj/release/idc_plugin_import.o  /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/platforms/libqwindows.a -ldwmapi -lwinspool -lwtsapi32 /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5EventDispatcherSupport.a -L/opt/mxe/usr/i686-w64-mingw32.static/lib /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5FontDatabaseSupport.a -ldwrite -ld2d1 /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5ThemeSupport.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5AccessibilitySupport.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5WindowsUIAutomationSupport.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqgif.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqicns.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqico.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqjp2.a -ljasper /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqjpeg.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqmng.a -lmng -llcms2 -lpthread /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqtga.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqtiff.a -ltiff -llzma -ljpeg /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqwbmp.a /opt/mxe/usr/i686-w64-mingw32.static/qt5/plugins/imageformats/libqwebp.a -lwebpmux -lwebpdemux -lwebp /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5Gui.a -ld3d11 -ldxgi -ldxguid -lharfbuzz -lcairo -lgobject-2.0 -lfontconfig -lfreetype -lm -lusp10 -lmsimg32 -lpixman-1 -lffi -lexpat -lbz2 -lpng16 -lharfbuzz_too -lfreetype_too -lglib-2.0 -lshlwapi -lpcre -lintl -liconv -lgdi32 -lcomdlg32 -loleaut32 -limm32 /opt/mxe/usr/i686-w64-mingw32.static/qt5/lib/libQt5Core.a -lmpr -luserenv -lversion -lz -lpcre2-16 -lzstd -lnetapi32 -lws2_32 -ladvapi32 -lkernel32 -lole32 -lshell32 -luuid -luser32 -lwinmm .obj/release/idc_resource_res.o   
/opt/mxe/usr/bin/i686-w64-mingw32.static-ld: /opt/mxe/usr/i686-w64-mingw32.static/lib/libjasper.a(jas_malloc.c.obj):jas_malloc.c:(.text+0x118b): undefined reference to `GetPhysicallyInstalledSystemMemory'
collect2: error: ld returned 1 exit status
make[6]: *** [Makefile.Release:69: ../../../bin/idc.exe] Error 1
make[6]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2/src/tools/idc'
make[5]: *** [Makefile:45: release] Error 2
make[5]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2/src/tools/idc'
make[4]: *** [Makefile:50: sub-idc-make_first] Error 2
make[4]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2/src/tools'
make[3]: *** [Makefile:94: sub-tools-make_first-ordered] Error 2
make[3]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2/src'
make[2]: *** [Makefile:53: sub-src-make_first] Error 2
make[2]: Leaving directory '/opt/mxe/tmp-qtactiveqt-i686-w64-mingw32.static/qtactiveqt-everywhere-src-5.15.2'
make[1]: *** [Makefile:881: build-only-qtactiveqt_i686-w64-mingw32.static] Error 2
make[1]: Leaving directory '/opt/mxe'
`

**Comment by MarcoBorrini99** [comment]
> MXE_TARGETS='i686-w64-mingw32.static' q

Docker can be used flawless to build mxe as personal experience, I'd suggest to start with an official "os container" such as ubuntu:latest or debian:latest and not this mysterious fffaraz/qt:mxe

**Comment by MarcoBorrini99** [comment]
I have to take a step back... indeed i686-w64-mingw32.static target is problematic and it's not a docker problem, same error here

**Comment by bluebrother** [comment]
I've ran into the same issue, but no docker, just Debian stable. In my case it breaks when trying to build qtdeclarative. After some digging:

1ee3c2e33ebc582b2ffc405b009b99530b798a8c updated jasper from 2.0.33 to 3.0.2. Apparently jasper now makes use of `GetPhysicallyInstalledSystemMemory`, which requires linking to [kernel32](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getphysicallyinstalledsystemmemory). Checking `usr/i686-w64-mingw32.static/lib/libkernel32.a` shows the symbol to be present in that lib, but the jasper pkg-config does not add kernel32 to the libs to link.

- Tried reverting the jasper part of the commit mentioned above: now things build fine. This is obviously a quick stop-gap solution.
- Tried changing `src/jasper.mk` to fixup ` jasper.pc` after build so `-lkernel32` is appended to the `Libs:` line. When building Qt jasper is used by the jpeg2000 qtimageformats plugin. With the changed `src/jasper.mk` now `qjp2.prl` includes kernel32 as lib (which it didn't do before). However, qtdeclarative still doesn't include -lkernel32 after -ljasper when calling the linker.
- Manually executed the failing linker call and added the missing -lkernel32, but it still fails. Not sure what I'm missing here -- adding the required lib to the linker command line should work, but I still get `mxe/usr/i686-w64-mingw32.static/lib/libjasper.a(jas_malloc.c.obj):jas_malloc.c:(.text+0x118b): undefined reference to 'GetPhysicallyInstalledSystemMemory'`

```
diff --git a/src/jasper.mk b/src/jasper.mk
index 0e53b3d5..901e0e5e 100644
--- a/src/jasper.mk
+++ b/src/jasper.mk
@@ -20,4 +20,9 @@ define $(PKG)_BUILD
         -DJAS_ENABLE_PROGRAMS=OFF
     $(MAKE) -C '$(BUILD_DIR)' -j '$(JOBS)'
     $(MAKE) -C '$(BUILD_DIR)' -j 1 install
+
+    # fix up Libs missing kernel32.
+    $(SED) -ie 's/Libs:\(.*\)/Libs: \1 -lkernel32/g' \
+           '$(PREFIX)/$(TARGET)/lib/pkgconfig/$(PKG).pc'
+
 endef
```

**Comment by MarcoBorrini99** [comment]
> I've ran into the same issue, but no docker, just Debian stable. In my case it breaks when trying to build qtdeclarative. After some digging:
> 
> [1ee3c2e](https://github.com/mxe/mxe/commit/1ee3c2e33ebc582b2ffc405b009b99530b798a8c) updated jasper from 2.0.33 to 3.0.2. Apparently jasper now makes use of `GetPhysicallyInstalledSystemMemory`, which requires linking to [kernel32](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getphysicallyinstalledsystemmemory). Checking `usr/i686-w64-mingw32.static/lib/libkernel32.a` shows the symbol to be present in that lib, but the jasper pkg-config does not add kernel32 to the libs to link.
> 
> * Tried reverting the jasper part of the commit mentioned above: now things build fine. This is obviously a quick stop-gap solution.
> * Tried changing `src/jasper.mk` to fixup ` jasper.pc` after build so `-lkernel32` is appended to the `Libs:` line. When building Qt jasper is used by the jpeg2000 qtimageformats plugin. With the changed `src/jasper.mk` now `qjp2.prl` includes kernel32 as lib (which it didn't do before). However, qtdeclarative still doesn't include -lkernel32 after -ljasper when calling the linker.
> * Manually executed the failing linker call and added the missing -lkernel32, but it still fails. Not sure what I'm missing here -- adding the required lib to the linker command line should work, but I still get `mxe/usr/i686-w64-mingw32.static/lib/libjasper.a(jas_malloc.c.obj):jas_malloc.c:(.text+0x118b): undefined reference to 'GetPhysicallyInstalledSystemMemory'`
>   **Edit:** `GetPhysicallyInstalledSystemMemory` requires at least Windows Vista, so this might be caused by targeting the wrong Windows version.
> 
> ```
> diff --git a/src/jasper.mk b/src/jasper.mk
> index 0e53b3d5..901e0e5e 100644
> --- a/src/jasper.mk
> +++ b/src/jasper.mk
> @@ -20,4 +20,9 @@ define $(PKG)_BUILD
>          -DJAS_ENABLE_PROGRAMS=OFF
>      $(MAKE) -C '$(BUILD_DIR)' -j '$(JOBS)'
>      $(MAKE) -C '$(BUILD_DIR)' -j 1 install
> +
> +    # fix up Libs missing kernel32.
> +    $(SED) -ie 's/Libs:\(.*\)/Libs: \1 -lkernel32/g' \
> +           '$(PREFIX)/$(TARGET)/lib/pkgconfig/$(PKG).pc'
> +
>  endef
> ```

Maybe settings `-D_WIN32_WINNT=0x0501` or `-DWINVER=0x0501` in a CFLAGS. Have to try...

**Comment by bluebrother** [comment]
> Maybe setting `-D_WIN32_WINNT=0x0501` or `-DWINVER=0x0501` in a CFLAGS. Have to try...

`-D_WIN32_WINNT=0x0601` is necessary. With that a simple test program compiles, with values less than 0x0601 it doesn't. This issue appears to be the same as in #2771 -- when trying to build libjasper shared I can get it link successfully when changing `CMakeLists.txt` to add that define to the target.

Looking at the preprocessor output when including `windows.h` `_WIN32_WINNT` is set to 0x0600 which is Vista. According to MSDN `GetPhysicallyInstalledSystemMemory` requires Vista with SP1, but 0x0601 is Win7. I assume that Vista SP1 is the same as Win7 in that regard.

So my initial thought of explicitly linking to kernel32 was wrong. Changing the default value for `_WIN32_WINNT` to 0x0601 in `usr/i686-w64-mingw32.shared/include/_mingw.h` makes things work. Which raises the question if the default Windows version should be increased to Win7 (i.e. 0x0601)

**Comment by rickvanosch** [comment]
@bluebrother thanks for the workaround, i will give this a try asap!

**Comment by tonytheodore** [comment]
> Which raises the question if the default Windows version should be increased to Win7 (i.e. 0x0601)

Should be fixed with #2785
